File: load-testing/test-100-jobs.py
#!/usr/bin/env python3

import requests
import time
import random
import sys
import json
from datetime import datetime

# Configuration
BASE_URL = sys.argv[1] if len(sys.argv) > 1 else 'http://kubs.local:50036'
JOB_TYPE = sys.argv[2] if len(sys.argv) > 2 else 'bcrypt'
TOTAL_JOBS = 100

# Colors for console output
class Colors:
    GREEN = '\033[32m'
    RED = '\033[31m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    RESET = '\033[0m'

def generate_job_payload(job_type, index):
    """Generate job payload based on type"""
    if job_type == 'prime':
        return {
            'type': 'prime',
            'payload': {
                'number': random.randint(1000, 50000),
                'complexity': random.randint(1, 10)
            }
        }
    elif job_type == 'bcrypt':
        return {
            'type': 'bcrypt',
            'payload': {
                'password': f'testpass{index}',
                'rounds': random.randint(8, 12)
            }
        }
    elif job_type == 'sort':
        algorithms = ['bubble', 'quick', 'merge']
        return {
            'type': 'sort',
            'payload': {
                'array': random.randint(100, 1000),
                'algorithm': random.choice(algorithms)
            }
        }
    else:
        raise ValueError(f'Invalid job type: {job_type}')

def submit_job(job_data, index):
    """Submit a single job"""
    try:
        start_time = time.time()
        response = requests.post(
            f'{BASE_URL}/api/jobs',
            json=job_data,
            headers={'Content-Type': 'application/json'},
            timeout=30
        )
        end_time = time.time()
        response_time = int((end_time - start_time) * 1000)
        
        if response.status_code == 201:
            job_id = response.json().get('jobId', 'Unknown')
            print(f'{Colors.GREEN}✓{Colors.RESET} Job {index}: {job_id} ({response_time}ms)')
            return True, job_id
        else:
            print(f'{Colors.RED}✗{Colors.RESET} Job {index}: Failed (HTTP {response.status_code}, {response_time}ms)')
            print(f'   Response: {response.text}')
            return False, None
    except Exception as e:
        print(f'{Colors.RED}✗{Colors.RESET} Job {index}: Error - {str(e)}')
        return False, None

def main():
    print(f'🚀 Submitting {TOTAL_JOBS} jobs to: {BASE_URL}')
    print(f'📝 Job type: {JOB_TYPE}')
    print('')
    
    success_count = 0
    failed_count = 0
    job_ids = []
    
    start_time = time.time()
    
    print(f'🔍 Testing {JOB_TYPE} jobs...')
    print('')
    
    # Submit all jobs
    for i in range(1, TOTAL_JOBS + 1):
        try:
            job_data = generate_job_payload(JOB_TYPE, i)
            success, job_id = submit_job(job_data, i)
            
            if success:
                success_count += 1
                job_ids.append(job_id)
            else:
                failed_count += 1
            
            # Small delay to avoid overwhelming the service
            time.sleep(0.1)
            
        except Exception as e:
            failed_count += 1
            print(f'{Colors.RED}✗{Colors.RESET} Job {i}: Error - {str(e)}')
    
    end_time = time.time()
    total_time = end_time - start_time
    
    print('')
    print('📊 Results:')
    print('===========')
    print(f'Total Jobs: {Colors.BLUE}{TOTAL_JOBS}{Colors.RESET}')
    print(f'Successful: {Colors.GREEN}{success_count}{Colors.RESET}')
    print(f'Failed: {Colors.RED}{failed_count}{Colors.RESET}')
    print(f'Success Rate: {Colors.GREEN}{round(success_count * 100 / TOTAL_JOBS)}%{Colors.RESET}')
    print(f'Total Time: {Colors.YELLOW}{total_time:.2f}s{Colors.RESET}')
    print(f'Average Time per Job: {Colors.YELLOW}{int(total_time * 1000 / TOTAL_JOBS)}ms{Colors.RESET}')
    
    if success_count == TOTAL_JOBS:
        print(f'\n{Colors.GREEN}🎉 All {TOTAL_JOBS} jobs submitted successfully!{Colors.RESET}')
    else:
        print(f'\n{Colors.YELLOW}⚠️  Some jobs failed. Check the service logs.{Colors.RESET}')
    
    print('')
    print('🔍 Check job status with:')
    if job_ids:
        print(f'   curl {BASE_URL}/api/jobs/{job_ids[0]}')
    print('📊 Monitor with Grafana dashboard')

if __name__ == '__main__':
    main()

File: load-testing/test-100-jobs.js
#!/usr/bin/env node

// Simple Node.js script to submit 100 jobs to the kubs submitter service
// Usage: node test-100-jobs.js [BASE_URL] [JOB_TYPE]

const http = require('http');
const https = require('https');

// Configuration
const BASE_URL = process.argv[2] || 'http://kubs.local:50036';
const JOB_TYPE = process.argv[3] || 'bcrypt';
const TOTAL_JOBS = 100;

// Colors for console output
const colors = {
    green: '\x1b[32m',
    red: '\x1b[31m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    reset: '\x1b[0m'
};

// Counters
let successCount = 0;
let failedCount = 0;
let jobIds = [];

console.log(`🚀 Submitting ${TOTAL_JOBS} jobs to: ${BASE_URL}`);
console.log(`📝 Job type: ${JOB_TYPE}`);
console.log('');

// Generate job payload based on type
function generateJobPayload(jobType, index) {
    switch (jobType) {
        case 'prime':
            return {
                type: 'prime',
                payload: {
                    number: Math.floor(Math.random() * 50000) + 1000,
                    complexity: Math.floor(Math.random() * 10) + 1
                }
            };
        case 'bcrypt':
            return {
                type: 'bcrypt',
                payload: {
                    password: `testpass${index}`,
                    rounds: Math.floor(Math.random() * 5) + 8
                }
            };
        case 'sort':
            const algorithms = ['bubble', 'quick', 'merge'];
            return {
                type: 'sort',
                payload: {
                    array: Math.floor(Math.random() * 1000) + 100,
                    algorithm: algorithms[Math.floor(Math.random() * algorithms.length)]
                }
            };
        default:
            throw new Error(`Invalid job type: ${jobType}`);
    }
}

// Submit a single job
function submitJob(jobData, index) {
    return new Promise((resolve) => {
        const startTime = Date.now();
        const url = new URL('/api/jobs', BASE_URL);
        
        const options = {
            hostname: url.hostname,
            port: url.port || (url.protocol === 'https:' ? 443 : 80),
            path: url.pathname,
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'User-Agent': 'Kubs-Test-Script/1.0'
            }
        };

        const client = url.protocol === 'https:' ? https : http;
        const req = client.request(options, (res) => {
            let responseData = '';
            res.on('data', chunk => responseData += chunk);
            res.on('end', () => {
                const endTime = Date.now();
                const responseTime = endTime - startTime;
                
                if (res.statusCode === 201) {
                    successCount++;
                    const jobId = JSON.parse(responseData).jobId;
                    jobIds.push(jobId);
                    console.log(`${colors.green}✓${colors.reset} Job ${index}: ${jobId} (${responseTime}ms)`);
                } else {
                    failedCount++;
                    console.log(`${colors.red}✗${colors.reset} Job ${index}: Failed (HTTP ${res.statusCode}, ${responseTime}ms)`);
                    console.log(`   Response: ${responseData}`);
                }
                resolve();
            });
        });

        req.on('error', (err) => {
            failedCount++;
            console.log(`${colors.red}✗${colors.reset} Job ${index}: Error - ${err.message}`);
            resolve();
        });

        req.write(JSON.stringify(jobData));
        req.end();
    });
}

// Main function
async function runTest() {
    const startTime = Date.now();
    
    console.log(`🔍 Testing ${JOB_TYPE} jobs...`);
    console.log('');
    
    // Submit all jobs
    for (let i = 1; i <= TOTAL_JOBS; i++) {
        try {
            const jobData = generateJobPayload(JOB_TYPE, i);
            await submitJob(jobData, i);
            
            // Small delay to avoid overwhelming the service
            await new Promise(resolve => setTimeout(resolve, 100));
        } catch (error) {
            failedCount++;
            console.log(`${colors.red}✗${colors.reset} Job ${i}: Error - ${error.message}`);
        }
    }
    
    const endTime = Date.now();
    const totalTime = endTime - startTime;
    
    console.log('');
    console.log('📊 Results:');
    console.log('===========');
    console.log(`Total Jobs: ${colors.blue}${TOTAL_JOBS}${colors.reset}`);
    console.log(`Successful: ${colors.green}${successCount}${colors.reset}`);
    console.log(`Failed: ${colors.red}${failedCount}${colors.reset}`);
    console.log(`Success Rate: ${colors.green}${Math.round(successCount * 100 / TOTAL_JOBS)}%${colors.reset}`);
    console.log(`Total Time: ${colors.yellow}${(totalTime / 1000).toFixed(2)}s${colors.reset}`);
    console.log(`Average Time per Job: ${colors.yellow}${Math.round(totalTime / TOTAL_JOBS)}ms${colors.reset}`);
    
    if (successCount === TOTAL_JOBS) {
        console.log(`\n${colors.green}🎉 All ${TOTAL_JOBS} jobs submitted successfully!${colors.reset}`);
    } else {
        console.log(`\n${colors.yellow}⚠️  Some jobs failed. Check the service logs.${colors.reset}`);
    }
    
    console.log('');
    console.log('🔍 Check job status with:');
    if (jobIds.length > 0) {
        console.log(`   curl ${BASE_URL}/api/jobs/${jobIds[0]}`);
    }
    console.log('📊 Monitor with Grafana dashboard');
}

// Run the test
runTest().catch(console.error);
